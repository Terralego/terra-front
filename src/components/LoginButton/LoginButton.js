import React, { useState, useEffect, useCallback } from 'react';
import classNames from 'classnames';
import {
  Tab,
  Tabs,
  Overlay,
  Classes,
  Button,
} from '@blueprintjs/core';
import { LoginForm, SignupForm } from '../../modules/Auth';
import withDeviceSize from '../../hoc/withDeviceSize';

import NavBarItemDesktop from '../NavBarItemDesktop';
import NavBarItemTablet from '../NavBarItemTablet';

export const LoginButton = ({
  authenticated,
  isMobileSized,
  isPhoneSized,
  logoutAction,
  t,
  ...props
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [hasCreated, setHasCreated] = useState(false);
  const [isNewlyAuthenticated, setIsNewlyAuthenticated] = useState(false);

  const wantLogout = isOpen && authenticated && !isNewlyAuthenticated;
  const NavBarItem = isMobileSized ? NavBarItemTablet : NavBarItemDesktop;

  useEffect(() => {
    if (authenticated) {
      setIsOpen(false);
    }
  }, [authenticated]);

  const confirmLogout = useCallback(() => {
    setIsOpen(false);
    logoutAction();
  }, [logoutAction]);

  const setOpen = useCallback(() => setIsOpen(true), []);
  const setClose = useCallback(() => setIsOpen(false), []);
  const onBeforeSubmitting = useCallback(() => setIsNewlyAuthenticated(true), []);
  const onAfterSubmitting = useCallback(() => setIsNewlyAuthenticated(false), []);
  const onCreate = useCallback(() => setHasCreated(true), []);

  return (
    <>
      <NavBarItem {...props} onClick={setOpen} />
      <Overlay
        className={classNames(
          Classes.OVERLAY_SCROLL_CONTAINER,
          Classes.DARK,
          'modal-signin',
        )}
        isOpen={isOpen}
        onClose={setClose}
      >
        <div
          className={classNames(
            Classes.CARD,
            Classes.ELEVATION_4,
            'modal-signin__form',
          )}
        >
          {isOpen && !authenticated
          && (
          <Tabs id="login">
            <Tab
              id="signin"
              title={t('auth.loginform.title')}
              panel={(
                <LoginForm
                  onBeforeSubmitting={onBeforeSubmitting}
                  onAfterSubmitting={onAfterSubmitting}
                  translate={t}
                />
              )}
            />
            <Tab
              id="signup"
              title={t('auth.signupform.title')}
              panel={hasCreated
                ? <p>{t('auth.signupform.done')}</p>
                : (
                  <SignupForm
                    translate={t}
                    showPassword={false}
                    onCreate={onCreate}
                  />
                )}
            />
          </Tabs>
          )}
          {wantLogout && (
          <div
            id="logout"
          >
            <p>{t('auth.logout.confirm.label')}</p>
            <div className="modal-signin__form-buttons">
              <Button
                text={t('auth.logout.cancel.button_label')}
                onClick={setClose}
                minimal
              />
              <Button
                text={t('auth.logout.confirm.button_label')}
                onClick={confirmLogout}
                intent="primary"
              />
            </div>
          </div>
          )}
        </div>
      </Overlay>
    </>
  );
};

export default withDeviceSize()(LoginButton);
