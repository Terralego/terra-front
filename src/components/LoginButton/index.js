import compose from '../../utils/compose';
import withDeviceSize from '../../hoc/withDeviceSize';
import { connectAuthProvider } from '../../modules/Auth';

import LoginButton from './LoginButton';

export default compose(
  withDeviceSize(),
  connectAuthProvider('authenticated', 'logoutAction'),
)(LoginButton);
