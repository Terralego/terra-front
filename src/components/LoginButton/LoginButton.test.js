import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import LoginButton from './LoginButton';

it('should render', () => {
  const { baseElement } = render(<LoginButton data-testid="login-button" />);
  fireEvent.click(baseElement.querySelector('[data-testid]'));
  expect(baseElement).toMatchSnapshot();
});


it('should render custom renderer when clicked', () => {
  const customRenderedText = 'Custom Renderer';
  const CustomRenderer = () => (<h1>{customRenderedText}</h1>);

  const { baseElement } = render(
    <LoginButton data-testid="login-button" render={CustomRenderer} />,
  );
  fireEvent.click(baseElement.querySelector('[data-testid]'));
  expect(screen.getByText(customRenderedText)).toBeDefined();
  expect(baseElement).toMatchSnapshot();
});
