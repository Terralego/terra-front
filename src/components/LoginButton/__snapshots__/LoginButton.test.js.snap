// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<body
  class="bp3-overlay-open"
>
  <div>
    <span
      class="bp3-popover-wrapper"
    >
      <span
        class="bp3-popover-target"
      >
        <span
          class="header-link"
          data-testid="login-button"
          tabindex="0"
        >
          <button
            class="bp3-button bp3-minimal header-button"
            type="button"
          />
        </span>
      </span>
    </span>
  </div>
  <div
    class="bp3-portal"
  >
    <div
      class="bp3-overlay bp3-overlay-open bp3-overlay-scroll-container bp3-dark modal-signin"
    >
      <div
        class="bp3-overlay-backdrop bp3-overlay-appear bp3-overlay-appear-active"
        tabindex="0"
      />
      <div
        class="bp3-card bp3-elevation-4 modal-signin__form bp3-overlay-content bp3-overlay-appear bp3-overlay-appear-active"
        tabindex="0"
      >
        <div
          class="bp3-tabs"
        >
          <div
            class="bp3-tab-list"
            role="tablist"
          >
            <div
              class="bp3-tab-indicator-wrapper"
              style="height: 0px; transform: translateX(0px) translateY(0px); width: 0px;"
            >
              <div
                class="bp3-tab-indicator"
              />
            </div>
            <div
              aria-controls="bp3-tab-panel_login_signin"
              aria-disabled="false"
              aria-expanded="true"
              aria-selected="true"
              class="bp3-tab"
              data-tab-id="signin"
              id="bp3-tab-title_login_signin"
              role="tab"
              tabindex="0"
            >
              Sign in
            </div>
          </div>
          <div
            aria-hidden="false"
            aria-labelledby="bp3-tab-title_login_signin"
            class="bp3-tab-panel"
            id="bp3-tab-panel_login_signin"
            role="tabpanel"
          >
            <div
              class="bp3-card bp3-elevation-0 login-form"
            >
              <form>
                <div
                  class="bp3-form-group"
                >
                  <label
                    class="bp3-label"
                    for="login"
                  >
                    Email
                     
                    <span
                      class="bp3-text-muted"
                    >
                      required
                    </span>
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="bp3-input-group"
                    >
                      <input
                        autocomplete="username"
                        class="bp3-input"
                        id="login"
                        placeholder="Email"
                        type="email"
                      />
                    </div>
                    <div
                      class="bp3-form-helper-text"
                    >
                      Type your email
                    </div>
                  </div>
                </div>
                <div
                  class="bp3-form-group"
                >
                  <label
                    class="bp3-label"
                    for="password"
                  >
                    Password
                     
                    <span
                      class="bp3-text-muted"
                    >
                      required
                    </span>
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="bp3-input-group"
                    >
                      <input
                        autocomplete="current-password"
                        class="bp3-input"
                        id="password"
                        placeholder="Password"
                        type="password"
                      />
                    </div>
                    <div
                      class="bp3-form-helper-text"
                    >
                      Type your password
                    </div>
                  </div>
                </div>
                <div
                  class="login-form__actions"
                >
                  <button
                    class="bp3-button"
                    type="submit"
                  >
                    <span
                      class="bp3-button-text"
                    >
                      Signin
                    </span>
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render custom renderer when clicked 1`] = `
<body
  class="bp3-overlay-open"
>
  <div>
    <span
      class="bp3-popover-wrapper"
    >
      <span
        class="bp3-popover-target"
      >
        <span
          class="header-link"
          data-testid="login-button"
          tabindex="0"
        >
          <button
            class="bp3-button bp3-minimal header-button"
            type="button"
          />
        </span>
      </span>
    </span>
  </div>
  <div
    class="bp3-portal"
  >
    <div
      class="bp3-overlay bp3-overlay-open bp3-overlay-scroll-container bp3-dark modal-signin"
    >
      <div
        class="bp3-overlay-backdrop bp3-overlay-appear bp3-overlay-appear-active"
        tabindex="0"
      />
      <div
        class="bp3-card bp3-elevation-4 modal-signin__form bp3-overlay-content bp3-overlay-appear bp3-overlay-appear-active"
        tabindex="0"
      >
        <div
          class="bp3-tabs"
        >
          <div
            class="bp3-tab-list"
            role="tablist"
          >
            <div
              class="bp3-tab-indicator-wrapper"
              style="height: 0px; transform: translateX(0px) translateY(0px); width: 0px;"
            >
              <div
                class="bp3-tab-indicator"
              />
            </div>
            <div
              aria-controls="bp3-tab-panel_login_signin"
              aria-disabled="false"
              aria-expanded="true"
              aria-selected="true"
              class="bp3-tab"
              data-tab-id="signin"
              id="bp3-tab-title_login_signin"
              role="tab"
              tabindex="0"
            >
              Sign in
            </div>
          </div>
          <div
            aria-hidden="false"
            aria-labelledby="bp3-tab-title_login_signin"
            class="bp3-tab-panel"
            id="bp3-tab-panel_login_signin"
            role="tabpanel"
          >
            <h1>
              Custom Renderer
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
